### PR TITLE
2119 create post consistent form input validation improvements

### PR DESCRIPTION
**Pull request by roschaefer**
## 🍰 Pullrequest
@ogerly I refactored quite a lot and I figured out how to connect individual form elements with `<ds-form>` (via `provide`) see below :point_down: 
### Issues

- None

### Todo

- [X] None

### CODE REVIEW

**Comment by ogerly** [comment]
it is the language that has to be selected to save a contribution. 
so it doesn't have to search for the global locale anymore. but the select default teaching for the language.

File: webapp/components/CategoriesSelect/CategoriesSelect.spec.js [hunk 1]
```diff
@@ -35,14 +37,19 @@ describe('CategoriesSelect.vue', () => {
     },
   ]
   beforeEach(() => {
+    provide = {
+      $parentForm: {
+        update: jest.fn(),
+      },
+    }
```
Comment (roschaefer) [inline webapp/components/CategoriesSelect/CategoriesSelect.spec.js]: This is how we can implement dedicated form input components, e.g. the `CategoriesSelect` component. @mattwr18

**Comment by alina-beck** [comment]
Hey, I haven't looked at the code yet and I'm a bit confused because there seem to be quite a few related PRs but I really liked the behavior and design @ogerly came up with in PR #2160! I hope it found (or will find) its way into this PR as well? :smile:

**Comment by datenbrei** [comment]
Yes. Also saw #2160 . I don't understand this one here. Looks for me like a discussion/decision not about ux, but about implementation details?!?

**Comment by roschaefer** [comment]
This is a PR into a PR. I did several refactorings and wanted to give @ogerly a sense what has changed.

I reopened https://github.com/Human-Connection/Human-Connection/pull/2160 so the changes here will find their way into `master`.

File: webapp/components/CategoriesSelect/CategoriesSelect.spec.js (hunk 1 above)
Comment (mattwr18) [reply webapp/components/CategoriesSelect/CategoriesSelect.spec.js]: looks pretty cool @roschaefer

File: backend/.env.template [hunk 2]
```diff
@@ -1,7 +1,6 @@
 NEO4J_URI=bolt://localhost:7687
 NEO4J_USERNAME=neo4j
 NEO4J_PASSWORD=letmein
-GRAPHQL_PORT=4000
 GRAPHQL_URI=http://localhost:4000
```
Comment (mattwr18) [inline backend/.env.template]: unrelated change?

File: backend/src/config/index.js [hunk 3]
```diff
@@ -16,7 +16,6 @@ const {
   NEO4J_URI = 'bolt://localhost:7687',
   NEO4J_USERNAME = 'neo4j',
   NEO4J_PASSWORD = 'neo4j',
-  GRAPHQL_PORT = 4000,
```
Comment (mattwr18) [inline backend/src/config/index.js]: oh, no code creep... :ghost:

File: backend/src/config/index.js [hunk 4]
```diff
@@ -36,7 +35,6 @@ export const smtpConfigs = {
 }
 export const neo4jConfigs = { NEO4J_URI, NEO4J_USERNAME, NEO4J_PASSWORD }
 export const serverConfigs = {
-  GRAPHQL_PORT,
```
Comment (mattwr18) [inline backend/src/config/index.js]: could you explain how these changes are related to the `PR`?

File: backend/src/index.js [hunk 5]
```diff
@@ -2,7 +2,8 @@ import createServer from './server'
 import CONFIG from './config'
 
 const { app } = createServer()
-app.listen({ port: CONFIG.GRAPHQL_PORT }, () => {
+const url = new URL(CONFIG.GRAPHQL_URI)
+app.listen({ port: url.port }, () => {
```
Comment (mattwr18) [inline backend/src/index.js]: seems unrelated to me...

File: cypress/integration/notifications/Mentions.feature [hunk 6]
```diff
@@ -20,6 +20,7 @@ Feature: Notification for a mention
       """
     And mention "@matt-rider" in the text
     And I select a category
+    And I choose a language for the post
```
Comment (mattwr18) [inline cypress/integration/notifications/Mentions.feature]: nice!

File: cypress/integration/post/WritePost.feature [hunk 7]
```diff
@@ -17,7 +17,8 @@ Feature: Create a post
       Human Connection is a free and open-source social network
       for active citizenship.
       """
-    Then I select a category
+    And I select a category
+    And I choose a language for the post
```
Comment (mattwr18) [inline cypress/integration/post/WritePost.feature]: yes, yes, yes

File: deployment/human-connection/templates/configmap.template.yaml [hunk 8]
```diff
@@ -4,7 +4,6 @@
   data:
     SMTP_HOST: "mailserver.human-connection"
     SMTP_PORT: "25"
-    GRAPHQL_PORT: "4000"
```
Comment (mattwr18) [inline deployment/human-connection/templates/configmap.template.yaml]: great clean up, but maybe a separate `PR` would be better?

File: docker-compose.maintenance.yml [hunk 9]
```diff
@@ -15,7 +15,6 @@ services:
     environment:
       - NEO4J_dbms_security_auth__enabled=false
       - NEO4J_dbms_memory_heap_max__size=2G
-      - GRAPHQL_PORT=4000
```
Comment (mattwr18) [inline docker-compose.maintenance.yml]: don't want to be a stickler, but...

File: package.json [hunk 10]
```diff
@@ -10,12 +10,11 @@
...
-    "cypress:backend:seeder": "cd backend && yarn run test:before:seeder",
-    "cypress:webapp": "cd webapp && cross-env GRAPHQL_URI=http://localhost:4123 yarn run dev",
-    "cypress:setup": "run-p cypress:backend:* cypress:webapp",
-    "cypress:run": "cypress run --browser chromium",
-    "cypress:open": "cypress open --browser chromium",
+    "cypress:backend": "cd backend && cross-env GRAPHQL_URI=http://localhost:4001 yarn run dev",
+    "cypress:webapp": "cd webapp && cross-env GRAPHQL_URI=http://localhost:4001 NUXT_PORT=3001 yarn run dev",
+    "cypress:setup": "run-p cypress:backend cypress:webapp",
+    "cypress:run": "cross-env CYPRESS_BASE_URL=http://localhost:3001 cypress run --browser chromium",
+    "cypress:open": "cross-env CYPRESS_BASE_URL=http://localhost:3001 cypress open --browser chromium",
```
Comment (mattwr18) [inline package.json]: is it because of this refactor? hard to tell how this is related to form validation

File: webapp/components/CategoriesSelect/CategoriesSelect.spec.js [hunk 11]
```diff
@@ -60,8 +67,8 @@ describe('CategoriesSelect.vue', () => {
         expect(wrapper.vm.selectedCategoryIds).toEqual([categories[0].id])
       })
 
-      it('emits an updateCategories event when the selectedCategoryIds changes', () => {
-        expect(wrapper.emitted().updateCategories[0][0]).toEqual([categories[0].id])
+      it('calls $parent.update with selected category ids', () => {
+        expect(provide.$parentForm.update).toHaveBeenCalledWith('categoryIds', ['cat9'])
```
Comment (mattwr18) [inline webapp/components/CategoriesSelect/CategoriesSelect.spec.js]: so we should refactor all our code like this?

File: webapp/components/CategoriesSelect/CategoriesSelect.vue [hunk 12]
```diff
@@ -28,16 +28,23 @@
...
   props: {
     existingCategoryIds: { type: Array, default: () => [] },
+    model: { type: String, required: true },
   },
   data() {
     return {
       categories: null,
       selectedMax: 3,
-      selectedCategoryIds: [],
+      selectedCategoryIds: this.existingCategoryIds,
```
Comment (mattwr18) [inline webapp/components/CategoriesSelect/CategoriesSelect.vue]: I thought I had seen that we don't have access to the store at the time this is read, but maybe I was wrong...

File: webapp/components/CategoriesSelect/CategoriesSelect.vue [hunk 13]
```diff
@@ -48,35 +55,15 @@ export default {
...
   methods: {
     toggleCategory(id) {
-      const index = this.selectedCategoryIds.indexOf(id)
-      if (index > -1) {
-        this.selectedCategoryIds.splice(index, 1)
-      } else {
-        this.selectedCategoryIds.push(id)
+      this.selectedCategoryIds = xor(this.selectedCategoryIds, [id])
+      if (this.$parentForm) {
+        this.$parentForm.update(this.model, this.selectedCategoryIds)
```
Comment (mattwr18) [inline webapp/components/CategoriesSelect/CategoriesSelect.vue]: really cool stuff!!

File: webapp/components/CategoriesSelect/CategoriesSelect.vue [hunk 14]
```diff
@@ -48,35 +55,15 @@ export default {
...
       }
     },
     isActive(id) {
-      const index = this.selectedCategoryIds.indexOf(id)
-      if (index > -1) {
-        return true
-      }
-      return false
+      return this.selectedCategoryIds.includes(id)
```
Comment (mattwr18) [inline webapp/components/CategoriesSelect/CategoriesSelect.vue]: of course!! so simple!!

File: webapp/components/ContributionForm/ContributionForm.spec.js [hunk 15]
```diff
@@ -109,81 +136,73 @@ describe('ContributionForm.vue', () => {
...
-        },
-      })
     })
 
     describe('CreatePost', () => {
       describe('language placeholder', () => {
-        it("displays the name that corresponds with the user's location code", () => {
+        it.skip("displays the name that corresponds with the user's location code", () => {
+          // Well not anymore right? We want the user to save the language
+          // excplicitly. I'll keep this test if we change our minds
```
Comment (mattwr18) [inline webapp/components/ContributionForm/ContributionForm.spec.js]: no, not anymore! I guess we could remove this...

File: webapp/components/ContributionForm/ContributionForm.spec.js [hunk 16]
```diff
@@ -109,81 +136,73 @@ describe('ContributionForm.vue', () => {
...
           expect(wrapper.find('.ds-select-placeholder').text()).toEqual('English')
         })
       })
 
       describe('invalid form submission', () => {
         it('title and content should not be empty ', async () => {
-          wrapper.find('.submit-button-for-test').trigger('click')
+          wrapper.find('form').trigger('submit')
```
Comment (mattwr18) [inline webapp/components/ContributionForm/ContributionForm.spec.js]: better

File: webapp/components/ContributionForm/ContributionForm.spec.js [hunk 17]
```diff
@@ -109,81 +136,73 @@ describe('ContributionForm.vue', () => {
...
           expect(mocks.$apollo.mutate).not.toHaveBeenCalled()
         })
 
         it('title should not be empty', async () => {
           await wrapper.vm.updateEditorContent(postContent)
-          wrapper.find('.submit-button-for-test').trigger('click')
+          wrapper.find('form').trigger('submit')
```
Comment (mattwr18) [inline webapp/components/ContributionForm/ContributionForm.spec.js]: like it

File: webapp/components/ContributionForm/ContributionForm.spec.js [hunk 18]
```diff
@@ -109,81 +136,73 @@ describe('ContributionForm.vue', () => {
...
           expect(mocks.$apollo.mutate).not.toHaveBeenCalled()
         })
 
         it('title should not be too long', async () => {
           postTitleInput = wrapper.find('.ds-input')
           postTitleInput.setValue(postTitleTooLong)
           await wrapper.vm.updateEditorContent(postContent)
-          wrapper.find('.submit-button-for-test').trigger('click')
+          wrapper.find('form').trigger('submit')
```
Comment (mattwr18) [inline webapp/components/ContributionForm/ContributionForm.spec.js]: same as :point_up:

File: webapp/components/ContributionForm/ContributionForm.spec.js [hunk 19]
```diff
@@ -109,81 +136,73 @@ describe('ContributionForm.vue', () => {
...
           expect(mocks.$apollo.mutate).not.toHaveBeenCalled()
         })
 
         it('title should not be too short', async () => {
           postTitleInput = wrapper.find('.ds-input')
           postTitleInput.setValue(postTitleTooShort)
           await wrapper.vm.updateEditorContent(postContent)
-          wrapper.find('.submit-button-for-test').trigger('click')
+          wrapper.find('form').trigger('submit')
```
Comment (mattwr18) [inline webapp/components/ContributionForm/ContributionForm.spec.js]: you get the point...

File: webapp/components/ContributionForm/ContributionForm.spec.js [hunk 20]
```diff
@@ -214,43 +233,48 @@ describe('ContributionForm.vue', () => {
...
+          wrapper.find(CategoriesSelect).setData({ categories })
+          wrapper
+            .findAll('li')
+            .at(1)
+            .trigger('click') // language
+          await wrapper
+            .find(CategoriesSelect)
+            .findAll('button')
+            .at(1)
+            .trigger('click')
```
Comment (mattwr18) [inline webapp/components/ContributionForm/ContributionForm.spec.js]: I think we can improve this! what if someone adds more buttons? it would be nice to do similar to what @Mogge did here https://github.com/Human-Connection/Human-Connection/blob/master/webapp/components/FilterPosts/FilterPosts.spec.js#L114 ,don't you think?

File: webapp/components/ContributionForm/ContributionForm.spec.js [hunk 21]
```diff
@@ -214,43 +233,48 @@ describe('ContributionForm.vue', () => {
...
         })
 
         it('supports changing the language', async () => {
           expectedParams.variables.language = 'de'
-          deutschOption = wrapper.findAll('li').at(0)
-          deutschOption.trigger('click')
-          wrapper.find('.submit-button-for-test').trigger('click')
+          wrapper
+            .findAll('li')
+            .at(0)
```
Comment (mattwr18) [inline webapp/components/ContributionForm/ContributionForm.spec.js]: same as :point_up_2:

File: webapp/components/ContributionForm/ContributionForm.spec.js [hunk 22]
```diff
@@ -275,11 +299,20 @@ describe('ContributionForm.vue', () => {
           postTitleInput.setValue(postTitle)
           await wrapper.vm.updateEditorContent(postContent)
           categoryIds = ['cat12']
-          wrapper.find(CategoriesSelect).vm.$emit('updateCategories', categoryIds)
+          wrapper.find(CategoriesSelect).setData({ categories })
+          wrapper
+            .findAll('li')
+            .at(1)
```
Comment (mattwr18) [inline webapp/components/ContributionForm/ContributionForm.spec.js]: same as :point_up_2:

File: webapp/components/ContributionForm/ContributionForm.spec.js [hunk 23]
```diff
@@ -275,11 +299,20 @@ describe('ContributionForm.vue', () => {
...
+            .trigger('click') // language
+          await wrapper
+            .find(CategoriesSelect)
+            .findAll('button')
+            .at(1)
```
Comment (mattwr18) [inline webapp/components/ContributionForm/ContributionForm.spec.js]: all of these, I think...
I like in enzyme where you have `filterWhere`, but you could just use `filter` from javascript as well... but the idea is that if someone adds a button the location will change, and all the tests will need to be updated, no?

File: webapp/components/ContributionForm/ContributionForm.spec.js [hunk 24]
```diff
@@ -356,19 +389,40 @@ describe('ContributionForm.vue', () => {
...
+          await wrapper
+            .find(CategoriesSelect)
+            .findAll('button')
+            .at(3)
+            .trigger('click')
+          await wrapper
+            .find(CategoriesSelect)
+            .findAll('button')
+            .at(4)
+            .trigger('click')
```
Comment (mattwr18) [inline webapp/components/ContributionForm/ContributionForm.spec.js]: maybe we don't need that many? don't know... I guess it's ok...

File: webapp/components/CategoriesSelect/CategoriesSelect.spec.js (hunk 11 above)
Comment (roschaefer) [reply webapp/components/CategoriesSelect/CategoriesSelect.spec.js]: Only custom form elements for `<ds-form>`

**Comment by mattwr18** [comment]
@roschaefer this doesn't address the component tests...? 
If you disagree with the code review, it would be nice to give reasons why, as I did when I asked for the changes

**Comment by roschaefer** [comment]
@mattwr18 no! This just got overlooked. I have troubles to understand what's going on in this test. Where do the categories of the components come from?
